# ANÁLISIS Y LIMPIEZA DE DATOS PARA TABLEAU

## FASE 0: PREPARACIÓN

In [59]:
# IMPORTAR LIBRERÍAS
import pandas as pd
import numpy as np

In [60]:
# CARGA DEL DATASET + COPIA DE SEGURIDAD + DIMENSIONES

# Cargar dataset
df = pd.read_csv(
    r"C:\Users\LORENA SIERRA\Downloads\amazon_cleaned_for_tableau.csv",
    low_memory=False
)
# Crear copia de trabajo
df_clean = df.copy()

#Dimensiones
print("=== DIMENSIONES GLOBALES ===")
print(f" Filas: {df_clean.shape[0]:,}")
print(f" Columnas: {df_clean.shape[1]}")
print(f" Memoria: {df_clean.memory_usage(deep=True).sum() / 1024**2:.1f} MB")

=== DIMENSIONES GLOBALES ===
 Filas: 123,839
 Columnas: 38
 Memoria: 184.4 MB


In [61]:
# MUESTRA DE TABLA

# Configurar pandas para mostrar todo
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', 15)  # Limitar ancho de texto

muestra_completa = df_clean.head(10)
display(muestra_completa)

index        order_id        date          status fulfilment sales_channel  \
0    1.0  171-9198151...  2022-04-30  Shipped - D...   Merchant     Amazon.in   
1    2.0  404-0687676...  2022-04-30         Shipped     Amazon     Amazon.in   
2    4.0  407-1069790...  2022-04-30         Shipped     Amazon     Amazon.in   
3    5.0  404-1490984...  2022-04-30         Shipped     Amazon     Amazon.in   
4    6.0  408-5748499...  2022-04-30         Shipped     Amazon     Amazon.in   
5    7.0  406-7807733...  2022-04-30  Shipped - D...   Merchant     Amazon.in   
6    8.0  407-5443024...  2022-04-30       Cancelled     Amazon     Amazon.in   
7    9.0  402-4393761...  2022-04-30         Shipped     Amazon     Amazon.in   
8   10.0  407-5633625...  2022-04-30         Shipped     Amazon     Amazon.in   
9   11.0  171-4638481...  2022-04-30         Shipped     Amazon     Amazon.in   

  ship_service_level   style             sku category size      asin  \
0        Standard     3781.0  JNE3781-KR-...    kurta  3XL    9332.0   
1       Expedited     3371.0   JNE3371-KR-XL    kurta   XL     744.0   
2       Expedited     3671.0  JNE3671-TU-...      Top  3XL   98714.0   
3       Expedited      264.0  SET264-KR-N...      Set   XL      87.0   
4       Expedited       95.0     J0095-SET-L      Set    L       8.0   
5        Standard     3405.0    JNE3405-KR-S    kurta    S    8144.0   
6       Expedited      200.0  SET200-KR-N...      Set  3XL     891.0   
7       Expedited     3461.0  JNE3461-KR-XXL    kurta  XXL     835.0   
8       Expedited     3160.0  JNE3160-KR-G-S    kurta    S     731.0   
9       Expedited     3500.0   JNE3500-KR-XS    kurta   XS  981173.0   

  courier_status  qty currency  amount    ship_city     ship_state  \
0        Shipped  1.0      INR   406.0    BENGALURU      KARNATAKA   
1        Shipped  1.0      INR   329.0  NAVI MUMBAI    MAHARASHTRA   
2        Shipped  1.0      INR   574.0      CHENNAI     TAMIL NADU   
3        Shipped  1.0      INR   824.0    GHAZIABAD  UTTAR PRADESH   
4        Shipped  1.0      INR   653.0   CHANDIGARH     CHANDIGARH   
5        Shipped  1.0      INR   399.0    HYDERABAD      TELANGANA   
6      Cancelled  0.0      NaN     NaN    HYDERABAD      TELANGANA   
7        Shipped  1.0      INR   363.0      Chennai     TAMIL NADU   
8        Shipped  1.0      INR   685.0      CHENNAI     TAMIL NADU   
9        Shipped  1.0      INR   364.0        NOIDA  UTTAR PRADESH   

   ship_postal_code ship_country   promotion_ids    b2b fulfilled_by  \
0        560085.0             IN  Amazon PLCC...  False    Easy Ship   
1        410210.0             IN  IN Core Fre...   True          NaN   
2        600073.0             IN             NaN  False          NaN   
3        201102.0             IN  IN Core Fre...  False          NaN   
4        160036.0             IN  IN Core Fre...  False          NaN   
5        500032.0             IN  Amazon PLCC...  False    Easy Ship   
6        500008.0             IN  IN Core Fre...  False          NaN   
7        600041.0             IN             NaN  False          NaN   
8        600073.0             IN             NaN  False          NaN   
9        201303.0             IN             NaN  False          NaN   

  unnamed_22  status_bucket  gross_revenue  order_year order_month  \
0        NaN      fulfilled          406.0        2022     2022-04   
1        NaN      fulfilled          329.0        2022     2022-04   
2        NaN      fulfilled          574.0        2022     2022-04   
3        NaN      fulfilled          824.0        2022     2022-04   
4        NaN      fulfilled          653.0        2022     2022-04   
5        NaN      fulfilled          399.0        2022     2022-04   
6        NaN  not_fulfilled            NaN        2022     2022-04   
7        NaN      fulfilled          363.0        2022     2022-04   
8        NaN      fulfilled          685.0        2022     2022-04   
9        NaN      fulfilled          364.0        2022

ESPECIFICACIONES DE LAS COLUMNAS

🔑IDs E IDENTIFICADORES
index: Índice

order_id: ID del pedido

sku: SKU (Unidad de Mantenimiento de Stock)

asin: Número de Identificación Estándar de Amazon

style: Estilo

📦 INFORMACIÓN DEL PEDIDO
date: Fecha

status: Estado del pedido

fulfilment: Cumplimiento

sales_channel: Canal de ventas

ship_service_level: Nivel de servicio de envío

courier_status: Estado del mensajero

status_bucket: Grupo de estado

🏷️ PRODUCTO
category: Categoría

size: Talla

💰 FINANZAS
currency: Moneda

amount: Monto

gross_revenue: Ingresos brutos

order_total: Total del pedido

aov_by_order: Valor promedio por pedido

is_negative_revenue: ¿Ingresos negativos?

📊 CANTIDADES
qty: Cantidad

order_line_count: Conteo de líneas del pedido

🌎 UBICACIÓN DE ENVÍO
ship_city: Ciudad de envío

ship_state: Estado de envío

ship_postal_code: Código postal de envío

ship_country: País de envío

🎯 MARKETING Y PROMOCIONES
promotion_ids: IDs de promociones

has_promo: ¿Tiene promoción?

👥 TIPO DE CLIENTE
b2b: Negocio a negocio

is_b2b: ¿Es negocio a negocio?

📅 FECHAS Y TIEMPOS
order_year: Año del pedido

order_month: Mes del pedido

order_week: Semana del pedido

order_weekday: Día de la semana del pedido

order_hour: Hora del pedido

is_weekend: ¿Es fin de semana?

❓ COLUMNAS PROBLEMÁTICAS
fulfilled_by: Cumplido por

unnamed_22: Sin nombre 22

## FASE 1: DIAGNÓSTICO INICIAL

### 1.1 TIPOS DE DATOS + VALORES UNICOS POR COLUMNA

In [62]:
# Crear DataFrame con tipos y valores únicos
tabla_completa = pd.DataFrame({
    'Columna': df_clean.columns,
    'Tipo_Dato': df_clean.dtypes.values,
    'Valores_Únicos': [df_clean[col].nunique() for col in df_clean.columns]
})

# Mostrar en tabla
display(tabla_completa)

print(f"\n RESUMEN: {len(tabla_completa)} columnas analizadas")

Columna Tipo_Dato  Valores_Únicos
0            index   float64          123839
1         order_id    object          115645
2             date    object              91
3           status    object              13
4       fulfilment    object               2
5    sales_channel    object               2
6   ship_servic...    object               2
7            style   float64             954
8              sku    object            7170
9         category    object               9
10            size    object              11
11            asin   float64            2255
12  courier_status    object               3
13             qty   float64              10
14        currency    object               1
15          amount   float64             867
16       ship_city    object            8761
17      ship_state    object              69
18  ship_postal...   float64            9342
19    ship_country    object               1
20   promotion_ids    object            5787
21             b2b      bool               2
22    fulfilled_by    object               1
23      unnamed_22    object               1
24   status_bucket    object               5
25   gross_revenue   float64             867
26      order_year     int64               1
27     order_month    object               4
28      order_week     int64              14
29   order_weekday    object               7
30      order_hour     int64               1
31      is_weekend      bool               2
32          is_b2b      bool               2
33       has_promo      bool               2
34     order_total   float64            2217
35  order_line_...     int64              12
36    aov_by_order   float64            2217
37  is_negative...      bool               1


 RESUMEN: 38 columnas analizadas


### 1.2 VALORES NULOS

In [63]:
# Solo columnas que tienen nulos
columnas_con_nulos = df_clean.columns[df_clean.isnull().sum() > 0]

print("CLASIFICACIÓN DE NULOS POR GRAVEDAD:\n")

# Contadores para el resumen
criticos = 0
altos = 0
medios = 0
leves = 0

for columna in columnas_con_nulos:
    nulos = df_clean[columna].isnull().sum()
    porcentaje = (nulos / len(df_clean)) * 100
    
    # Clasificar por gravedad
    if porcentaje > 50:
        gravedad = "🔴 CRÍTICO"
        criticos += 1
    elif porcentaje > 10:
        gravedad = "🟡 ALTO"
        altos += 1
    elif porcentaje > 1:
        gravedad = "🟠 MEDIO" 
        medios += 1
    else:
        gravedad = "🟢 LEVE"
        leves += 1
    
    print(f"{gravedad} {columna}:")
    print(f"   • {nulos} nulos ({porcentaje:.1f}%)")

print(f"\n RESUMEN DE GRAVEDAD:")
print(f"   🔴 CRÍTICOS: {criticos} columnas (>50% nulos)")
print(f"   🟡 ALTOS: {altos} columnas (10-50% nulos)")
print(f"   🟠 MEDIOS: {medios} columnas (1-10% nulos)") 
print(f"   🟢 LEVES: {leves} columnas (<1% nulos)")
print(f"    TOTAL: {len(columnas_con_nulos)} columnas con nulos")

CLASIFICACIÓN DE NULOS POR GRAVEDAD:

🟠 MEDIO courier_status:
   • 1736 nulos (1.4%)
🟠 MEDIO currency:
   • 7795 nulos (6.3%)
🟠 MEDIO amount:
   • 7795 nulos (6.3%)
🟢 LEVE ship_city:
   • 30 nulos (0.0%)
🟢 LEVE ship_state:
   • 30 nulos (0.0%)
🟢 LEVE ship_postal_code:
   • 30 nulos (0.0%)
🟢 LEVE ship_country:
   • 30 nulos (0.0%)
🟡 ALTO promotion_ids:
   • 44017 nulos (35.5%)
🔴 CRÍTICO fulfilled_by:
   • 89698 nulos (72.4%)
🟡 ALTO unnamed_22:
   • 46921 nulos (37.9%)
🟠 MEDIO gross_revenue:
   • 7795 nulos (6.3%)
🟠 MEDIO order_total:
   • 7685 nulos (6.2%)
🟠 MEDIO aov_by_order:
   • 7685 nulos (6.2%)

 RESUMEN DE GRAVEDAD:
   🔴 CRÍTICOS: 1 columnas (>50% nulos)
   🟡 ALTOS: 2 columnas (10-50% nulos)
   🟠 MEDIOS: 6 columnas (1-10% nulos)
   🟢 LEVES: 4 columnas (<1% nulos)
    TOTAL: 13 columnas con nulos


### 1.3 DETECCIÓN DE REDUNDANCIAS

In [64]:
redundancias = []
constantes = []

# 1. Verificar columnas que podrían ser duplicadas
print(" POSIBLES REDUNDANCIAS:")
if 'b2b' in df_clean.columns and 'is_b2b' in df_clean.columns:
    son_iguales = (df_clean['b2b'] == df_clean['is_b2b']).all()
    print(f"   • b2b vs is_b2b: ¿Son iguales? {son_iguales}")
    if son_iguales:
        redundancias.append(('b2b', 'is_b2b', 'Columnas idénticas'))

# 2. Buscar columnas constantes (un solo valor)
print("\n COLUMNAS CONSTANTES:")
for columna in df_clean.columns:
    valores_unicos = df_clean[columna].nunique()
    if valores_unicos == 1:
        valor = df_clean[columna].iloc[0]
        print(f"   • {columna}: Solo tiene un valor -> {valor}")
        constantes.append((columna, valor))

# 3. Verificar columnas redundantes por significado
print("\n COLUMNAS POTENCIALMENTE INNECESARIAS:")
if 'index' in df_clean.columns:
    print("   • index: Redundante con índice automático de pandas")
    redundancias.append(('index', 'índice_automático', 'Duplicado del índice'))

if 'currency' in df_clean.columns:
    valores_currency = df_clean['currency'].value_counts()
    if len(valores_currency) == 1:
        print(f"   • currency: Solo tiene un valor -> {valores_currency.index[0]}")
        constantes.append(('currency', valores_currency.index[0]))

if 'ship_country' in df_clean.columns:
    valores_pais = df_clean['ship_country'].value_counts()
    if len(valores_pais) == 1:
        print(f"   • ship_country: Solo tiene un valor -> {valores_pais.index[0]}")
        constantes.append(('ship_country', valores_pais.index[0]))

print(f"\n RESUMEN REDUNDANCIAS:")
print(f"   • Columnas redundantes identificadas: {len(redundancias)}")
print(f"   • Columnas constantes identificadas: {len(constantes)}")

 POSIBLES REDUNDANCIAS:
   • b2b vs is_b2b: ¿Son iguales? True

 COLUMNAS CONSTANTES:
   • currency: Solo tiene un valor -> INR
   • ship_country: Solo tiene un valor -> IN
   • fulfilled_by: Solo tiene un valor -> Easy Ship
   • unnamed_22: Solo tiene un valor -> nan
   • order_year: Solo tiene un valor -> 2022
   • order_hour: Solo tiene un valor -> 0
   • is_negative_revenue: Solo tiene un valor -> False

 COLUMNAS POTENCIALMENTE INNECESARIAS:
   • index: Redundante con índice automático de pandas
   • currency: Solo tiene un valor -> INR
   • ship_country: Solo tiene un valor -> IN

 RESUMEN REDUNDANCIAS:
   • Columnas redundantes identificadas: 2
   • Columnas constantes identificadas: 9


### 1.4 HALLAZGOS PRINCIPALES IDENTIFICADOS

PROBLEMAS CRÍTICOS (1.1)
fulfilled_by: 72.4% nulos 🔴 + 1 valor único

unnamed_22: 37.9% nulos 🟡 + 1 valor único

 PROBLEMAS DE TIPOS (1.2)
date: object (debería ser datetime)
qty: float64 (debería ser int64)
Varias categorías como object (podrían optimizarse)

 REDUNDANCIAS IDENTIFICADAS (1.3)
b2b vs is_b2b: Columnas idénticas (ambas con 2 valores únicos)
index: Redundante con índice automático
currency: Valor constante 'INR' (1 valor único)
ship_country: Valor constante 'IN' (1 valor único)
is_negative_revenue: Valor constante 'False' (1 valor único)
order_year: Valor constante '2022' (1 valor único)
order_hour: Valor constante '0' (1 valor único)

MÉTRICAS FINALES
Total columnas analizadas: 38
Columnas con nulos: 13
Columnas redundantes/constantes: 9
Columnas con tipos a mejorar: 3+

## FASE 2: LIMPIEZA

EXPLICACIÓN DETALLADA DE CADA PROBLEMA

*** b2b vs is_b2b: Columnas idénticas
PROBLEMA: Información duplicada
Ambas columnas contienen EXACTAMENTE la misma información
Ambas tienen 2 valores únicos: True/False
Ambas miden si la venta es Business-to-Business
Ocupan espacio innecesario en la base de datos
SOLUCIÓN: Mantener solo una (is_b2b es más descriptiva)

*** index: Redundante con índice automático
PROBLEMA: Columna innecesaria
Es un duplicado del índice automático que pandas ya crea (0, 1, 2, 3...)
No agrega información de negocio
OcUpa memoria sin aportar valor
SOLUCIÓN: Eliminar - ya tenemos indexación automática

*** currency: Valor constante 'INR'
PROBLEMA: Información redundante
Solo tiene un valor: 'INR' (Rupia India)
Todas las transacciones son en la misma moneda
No ayuda en segmentación o filtros
En Tableau, podemos poner 'INR' como etiqueta fija
SOLUCIÓN: Eliminar - información constante

*** ship_country: Valor constante 'IN'
PROBLEMA: Información redundante
Solo tiene un valor: 'IN' (India)
Todas las ventas son dentro de India
No permite análisis geográfico comparativo
Ocupa espacio sin variabilidad
SOLUCIÓN: Eliminar - todos los envíos son nacionales

*** is_negative_revenue: Valor constante 'False'
PROBLEMA: Métrica inútil
Solo tiene un valor: False
Indica que NINGUNA venta tuvo ingresos negativos
No aporta información para el análisis
SOLUCIÓN: Eliminar - no hay casos de ingresos negativos

*** order_year: Valor constante '2022'
PROBLEMA: Información redundante
Solo tiene un valor: 2022
Todos los datos son del mismo año
Ya tenemos la columna 'date' para análisis temporal
SOLUCIÓN: Eliminar - información constante y redundante

*** order_hour: Valor constante '0'
PROBLEMA: Datos incorrectos o sin variabilidad
Solo tiene un valor: 0 (medianoche)
Es improbable que TODAS las ventas sean exactamente a la misma hora
Sugiere problema en la captura de datos
SOLUCIÓN: Eliminar - datos probablemente erróneos

PROBLEMAS CRÍTICOS DE NULOS
🔴 fulfilled_by: 72.4% nulos
PROBLEMA: Columna casi vacía
72% de los datos faltan
Solo el 28% tiene valor ("Easy Ship")
Demasiados nulos para ser útil en análisis
SOLUCIÓN: Eliminar - demasiada información perdida

🟡 unnamed_22: 37.9% nulos
PROBLEMA: Columna misteriosa y con muchos nulos
38% de datos faltantes
No tiene nombre descriptivo
Cuando tiene valor, solo es "False"
SOLUCIÓN: Eliminar - sin significado + muchos nulos

🟠 promotion_ids: 35.5% nulos
PROBLEMA: Muchos nulos pero potencialmente útil
35% de datos faltantes
Los nulos significan "sin promoción"
Podría ser útil para análisis de marketing
SOLUCIÓN: Mantener pero rellenar nulos con "Sin promoción"

🟠 amount/gross_revenue: 6.3% nulos
PROBLEMA: Nulos en datos financieros críticos
6% de órdenes sin monto
Probablemente son órdenes canceladas
SOLUCIÓN: Investigar y decidir (eliminar o marcar como canceladas)

RESUMEN COMPLETO:
ELIMINAR POR NULOS:
fulfilled_by (72% nulos)
unnamed_22 (38% nulos)

ELIMINAR POR REDUNDANCIA/CONSTANTES:
index, currency, ship_country, is_negative_revenue, order_year, order_hour, b2b

MANTENER Y ARREGLAR:
promotion_ids (rellenar nulos)
amount/gross_revenue (investigar nulos)

### 2.1 ELIMINACIÓN POR NULOS

In [65]:
# Columnas con muchos nulos (basado en diagnóstico FASE 1)
columnas_nulos_criticos = [
    'fulfilled_by',    # 🔴 72.4% nulos - demasiado perdido
    'unnamed_22'       # 🟡 37.9% nulos + sin significado
]

print(" ELIMINANDO COLUMNAS CON MUCHOS NULOS:")
for columna in columnas_nulos_criticos:
    if columna in df_clean.columns:
        nulos = df_clean[columna].isnull().sum()
        print(f"   • {columna}: {nulos} nulos ({(nulos/len(df_clean)*100):.1f}%)")

print(f"\n ANTES: {df_clean.shape[1]} columnas")

# Ejecutar eliminación
df_clean = df_clean.drop(columns=columnas_nulos_criticos, errors='ignore')

print(f" DESPUÉS: {df_clean.shape[1]} columnas")
print(f" Eliminadas: {len(columnas_nulos_criticos)} columnas con muchos nulos")

 ELIMINANDO COLUMNAS CON MUCHOS NULOS:
   • fulfilled_by: 89698 nulos (72.4%)
   • unnamed_22: 46921 nulos (37.9%)

 ANTES: 38 columnas
 DESPUÉS: 36 columnas
 Eliminadas: 2 columnas con muchos nulos


### 2.2: ELIMINACIÓN POR REDUNDANCIAS Y CONSTANTES

In [66]:
# Columnas redundantes y constantes
columnas_redundantes = [
    'index', 'currency', 'ship_country', 'is_negative_revenue', 
    'order_year', 'order_hour', 'b2b'
]

print(f" ANTES: {df_clean.shape[1]} columnas")

# Ejecutar eliminación
df_clean = df_clean.drop(columns=columnas_redundantes, errors='ignore')

print(f" DESPUÉS: {df_clean.shape[1]} columnas")
print(f" Eliminadas: {len(columnas_redundantes)} columnas redundantes/constantes")


 ANTES: 36 columnas
 DESPUÉS: 29 columnas
 Eliminadas: 7 columnas redundantes/constantes


### 2.3 MANEJO DE NULOS EN COLUMNAS IMPORTANTES

In [67]:
# 1. Rellenar nulos en promotion_ids
print("1.  RELLENANDO PROMOTION_IDS:")
nulos_promo = df_clean['promotion_ids'].isnull().sum()
print(f"   • Nulos antes: {nulos_promo}")

df_clean['promotion_ids'] = df_clean['promotion_ids'].fillna('Sin promoción')

nulos_promo_despues = df_clean['promotion_ids'].isnull().sum()
print(f"   •Nulos después: {nulos_promo_despues}")
print(f"    Nulos arreglados: {nulos_promo}")

# 2. Investigar nulos en amount/gross_revenue
print("\n2.  INVESTIGANDO NULOS EN AMOUNT:")
nulos_amount = df_clean['amount'].isnull().sum()
print(f"   • Nulos en amount: {nulos_amount}")

if nulos_amount > 0:
    # Verificar estados de las órdenes sin amount
    ordenes_sin_amount = df_clean[df_clean['amount'].isnull()]
    print(f"   • Estados de órdenes sin amount:")
    estados = ordenes_sin_amount['status'].value_counts()
    for estado, cantidad in estados.items():
        print(f"     - {estado}: {cantidad} órdenes")

print(f"\n RESUMEN FASE 2.3:")
print(f"   • Promotion_ids: {nulos_promo} nulos → 0 nulos")
print(f"   • Amount: {nulos_amount} nulos por investigar")

1.  RELLENANDO PROMOTION_IDS:
   • Nulos antes: 44017
   •Nulos después: 0
    Nulos arreglados: 44017

2.  INVESTIGANDO NULOS EN AMOUNT:
   • Nulos en amount: 7795
   • Estados de órdenes sin amount:
     - Cancelled: 7566 órdenes
     - Shipped: 208 órdenes
     - Shipped - Delivered to Buyer: 8 órdenes
     - Shipping: 8 órdenes
     - Shipped - Returned to Seller: 3 órdenes
     - Pending: 2 órdenes

 RESUMEN FASE 2.3:
   • Promotion_ids: 44017 nulos → 0 nulos
   • Amount: 7795 nulos por investigar


### 2.4 MANEJO DE ÓRDENES SIN MONTO + DOCUMENTACIÓN

In [69]:
# Crear columna para identificar cancelaciones
print("1.  CREANDO COLUMNA DE IDENTIFICACIÓN:")
df_clean['is_cancelled'] = df_clean['status'] == 'Cancelled'
cancelaciones = df_clean['is_cancelled'].sum()
print(f"   • Órdenes canceladas identificadas: {cancelaciones}")

# 2. Para las canceladas, establecer amount = 0
print("\n2.  AJUSTANDO MONTOS DE CANCELADAS:")
ordenes_sin_monto = df_clean['amount'].isnull().sum()
print(f"   • Órdenes sin monto antes: {ordenes_sin_monto}")

# Poner amount = 0 solo para canceladas
df_clean.loc[df_clean['is_cancelled'], 'amount'] = 0
df_clean.loc[df_clean['is_cancelled'], 'gross_revenue'] = 0

ordenes_sin_monto_despues = df_clean['amount'].isnull().sum()
print(f"   • Órdenes sin monto después: {ordenes_sin_monto_despues}")
print(f"    Montos ajustados en {cancelaciones} órdenes canceladas")

1.  CREANDO COLUMNA DE IDENTIFICACIÓN:
   • Órdenes canceladas identificadas: 13196

2.  AJUSTANDO MONTOS DE CANCELADAS:
   • Órdenes sin monto antes: 7795
   • Órdenes sin monto después: 229
    Montos ajustados en 13196 órdenes canceladas


In [74]:
# ANÁLISIS DETALLADO DE ÓRDENES ENVIADAS SIN MONTO

ordenes_problematicas = df_clean[(df_clean['amount'].isnull()) & (~df_clean['is_cancelled'])]

print(" HALLAZGOS CRÍTICOS:")

# 1. Verificar si tienen otros datos de venta
print("\n1.  DATOS FINANCIEROS RELACIONADOS:")
print(f"   • Tienen order_total: {ordenes_problematicas['order_total'].notna().sum()}")
print(f"   • Tienen gross_revenue: {ordenes_problematicas['gross_revenue'].notna().sum()}")
print(f"   • Tienen aov_by_order: {ordenes_problematicas['aov_by_order'].notna().sum()}")

# 2. Verificar si son del mismo canal o categoría
print("\n2.  PATRONES COMUNES:")
print(f"   • Canal de ventas: {ordenes_problematicas['sales_channel'].value_counts().to_dict()}")
print(f"   • Categorías: {ordenes_problematicas['category'].value_counts().to_dict()}")

# 3. Verificar si tienen cantidad (qty)
print(f"\n3.  INFORMACIÓN DE PRODUCTO:")
print(f"   • Tienen cantidad (qty): {ordenes_problematicas['qty'].notna().sum()}")
if ordenes_problematicas['qty'].notna().any():
    print(f"   • Cantidades: {ordenes_problematicas['qty'].value_counts().to_dict()}")

# 4. Verificar si son órdenes con promociones especiales
print(f"\n4.  PROMOCIONES:")
print(f"   • Tienen promotion_ids: {ordenes_problematicas['promotion_ids'].notna().sum()}")

 HALLAZGOS CRÍTICOS:

1.  DATOS FINANCIEROS RELACIONADOS:
   • Tienen order_total: 106
   • Tienen gross_revenue: 0
   • Tienen aov_by_order: 106

2.  PATRONES COMUNES:
   • Canal de ventas: {'Non-Amazon': 123, 'Amazon.in': 106}
   • Categorías: {'Set': 104, 'kurta': 100, 'Top': 10, 'Western Dress': 9, 'Blouse': 4, 'Bottom': 1, 'Ethnic Dress': 1}

3.  INFORMACIÓN DE PRODUCTO:
   • Tienen cantidad (qty): 229
   • Cantidades: {1.0: 114, 0.0: 106, 2.0: 4, 15.0: 1, 9.0: 1, 13.0: 1, 4.0: 1, 3.0: 1}

4.  PROMOCIONES:
   • Tienen promotion_ids: 229


### 2.5 ELIMINACIÓN DE LAS 229 ÓRDENES PROBLEMÁTICAS

In [76]:
# Registrar estado antes de la eliminación
filas_antes = df_clean.shape[0]
nulos_antes = df_clean['amount'].isnull().sum()

print(" ESTADO ANTES:")
print(f"   • Filas totales: {filas_antes:,}")
print(f"   • Órdenes sin amount: {nulos_antes}")
print(f"   • Porcentaje problemático: {(nulos_antes/filas_antes*100):.2f}%")

# Eliminar órdenes sin amount (las 229 problemáticas)
df_clean = df_clean.dropna(subset=['amount'])

# Registrar estado después
filas_despues = df_clean.shape[0]
nulos_despues = df_clean['amount'].isnull().sum()

 ESTADO ANTES:
   • Filas totales: 123,839
   • Órdenes sin amount: 0
   • Porcentaje problemático: 0.00%


## FASE 3: OPTIMIZACIÓN DE TIPOS DE DATOS

In [78]:
# 1. Fecha → datetime (CRÍTICO para Tableau)
print("1.  CONVIRTIENDO FECHA:")
print(f"   • Antes: {df_clean['date'].dtype}")
df_clean['date'] = pd.to_datetime(df_clean['date'])
print(f"   • Después: {df_clean['date'].dtype}")

# 2. Cantidad → entero
print(f"\n2.  CONVIRTIENDO CANTIDAD:")
print(f"   • Antes: {df_clean['qty'].dtype}")
df_clean['qty'] = df_clean['qty'].astype('int64')
print(f"   • Después: {df_clean['qty'].dtype}")

# 3. Optimizar categorías principales
print(f"\n3.  OPTIMIZANDO CATEGORÍAS:")
categorias_optimizar = ['category', 'size', 'status', 'fulfilment']
for col in categorias_optimizar:
    if col in df_clean.columns:
        print(f"   • {col}: object → category")
        df_clean[col] = df_clean[col].astype('category')


1.  CONVIRTIENDO FECHA:
   • Antes: datetime64[ns]
   • Después: datetime64[ns]

2.  CONVIRTIENDO CANTIDAD:
   • Antes: int64
   • Después: int64

3.  OPTIMIZANDO CATEGORÍAS:
   • category: object → category
   • size: object → category
   • status: object → category
   • fulfilment: object → category


## FASE 4: EXPORTACIÓN FINAL

In [79]:
# Guardar dataset limpio
archivo_final = 'amazon_sales_clean_tableau.csv'
df_clean.to_csv(archivo_final, index=False)

print(f" DATASET EXPORTADO: {archivo_final}")
print(f" RESUMEN FINAL:")
print(f"   • Filas: {df_clean.shape[0]:,}")
print(f"   • Columnas: {df_clean.shape[1]}")
print(f"   • Tamaño: {df_clean.memory_usage(deep=True).sum() / 1024**2:.1f} MB")

 DATASET EXPORTADO: amazon_sales_clean_tableau.csv
 RESUMEN FINAL:
   • Filas: 123,839
   • Columnas: 30
   • Tamaño: 130.9 MB


In [82]:
# Dónde conseguir el archivo guardado?
import os
print(" CARPETA ACTUAL:")
print(os.getcwd())
print(f"\n ARCHIVO CSV:")
print("amazon_sales_clean_tableau.csv")

 CARPETA ACTUAL:
C:\Users\LORENA SIERRA\3D Objects

 ARCHIVO CSV:
amazon_sales_clean_tableau.csv
